keadm/cmd/keadm/app/cmd/edge/upgrade.go

```diff
@@ -0,0 +1,397 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package edge
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/spf13/cobra"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/yaml"
+
+	"github.com/kubeedge/kubeedge/common/constants"
+	commontypes "github.com/kubeedge/kubeedge/common/types"
+	"github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/common"
+	"github.com/kubeedge/kubeedge/keadm/cmd/keadm/app/cmd/util"
+	"github.com/kubeedge/kubeedge/pkg/apis/componentconfig/edgecore/v1alpha1"
+	upgradev1alpha1 "github.com/kubeedge/kubeedge/pkg/apis/upgrade/v1alpha1"
+	"github.com/kubeedge/kubeedge/pkg/image"
+)
+
+// NewEdgeUpgrade returns KubeEdge edge upgrade command.
+func NewEdgeUpgrade() *cobra.Command {
+	upgradeOptions := newUpgradeOptions()
+
+	cmd := &cobra.Command{
+		Use:   "upgrade",
+		Short: "Upgrade edge component. Upgrade the edge node to the desired version.",
+		Long:  "Upgrade edge component. Upgrade the edge node to the desired version.",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// upgrade edgecore
+			return upgradeOptions.upgrade()
+		},
+	}
+
+	addUpgradeFlags(cmd, upgradeOptions)
+	return cmd
+}
+
+// newJoinOptions returns a struct ready for being used for creating cmd join flags.
+func newUpgradeOptions() *UpgradeOptions {
+	opts := &UpgradeOptions{}
+	opts.ToVersion = "v" + common.DefaultKubeEdgeVersion
+	opts.Config = constants.DefaultConfigDir + "edgecore.yaml"
+
+	return opts
+}
+
+func (up *UpgradeOptions) upgrade() error {
+	// get EdgeCore configuration from edgecore.yaml config file
+	data, err := os.ReadFile(up.Config)
+	if err != nil {
+		return fmt.Errorf("failed to read config file %s: %v", up.Config, err)
+	}
+
+	configure := &v1alpha1.EdgeCoreConfig{}
+	err = yaml.Unmarshal(data, configure)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal config file %s: %v", up.Config, err)
+	}
+
+	upgrade := Upgrade{
+		UpgradeID:      up.UpgradeID,
+		HistoryID:      up.HistoryID,
+		FromVersion:    up.FromVersion,
+		ToVersion:      up.ToVersion,
+		Repository:     up.Repository,
+		ConfigFilePath: up.Config,
+		EdgeCoreConfig: configure,
+	}
+
+	defer func() {
+		// report upgrade result to cloudhub
+		if err := upgrade.reportUpgradeResult(); err != nil {
+			klog.Errorf("failed to report upgrade result to cloud: %v", err)
+		}
+		// cleanup idempotency record
+		if err := os.Remove(filepath.Join(util.KubeEdgePath, "idempotency_record")); err != nil {
```

Comment:
use a const for idempotency_record and add comments for the file

Reply:
fixed